File: android/app/src/main/kotlin/hk/smsguard/app/rules/SsrsVerifier.kt
package hk.smsguard.app.rules

private val HASH_PREFIX_PATTERN = Regex("^#([A-Za-z0-9_]+)\\b")

sealed class SsrsCheckOutcome {
    data class TrustedRegisteredPrefix(val observedPrefix: String) : SsrsCheckOutcome()
    data class PhishingClaimsOrgWithoutPrefix(val matchedOrg: OrgPrefixMapping) : SsrsCheckOutcome()
    data class UnknownHashPrefix(val observedPrefix: String) : SsrsCheckOutcome()
    data object NoSignal : SsrsCheckOutcome()
}

fun checkSsrs(sms: IncomingSms, registry: SsrsRegistry): SsrsCheckOutcome {
    val observedPrefix = extractHashPrefix(sms.senderId)

    if (observedPrefix != null) {
        val isRegistered = registry.registeredPrefixes.any {
            it.prefix.equals(observedPrefix, ignoreCase = true)
        }
        return if (isRegistered) {
            SsrsCheckOutcome.TrustedRegisteredPrefix(observedPrefix)
        } else {
            SsrsCheckOutcome.UnknownHashPrefix(observedPrefix)
        }
    }

    val matchedOrg = findOrgClaimedInBody(sms.body, registry.orgToPrefix)
    return if (matchedOrg != null) {
        SsrsCheckOutcome.PhishingClaimsOrgWithoutPrefix(matchedOrg)
    } else {
        SsrsCheckOutcome.NoSignal
    }
}

fun ssrsCheckToVerdict(outcome: SsrsCheckOutcome): Verdict? = when (outcome) {
    is SsrsCheckOutcome.TrustedRegisteredPrefix -> Verdict(
        label = VerdictLabel.TRUSTED,
        score = 1.0,
        firedRuleIds = listOf("ssrs.trusted_registered_prefix"),
        explanationKey = "ssrs.trusted",
        explanationParams = mapOf("prefix" to outcome.observedPrefix),
    )
    // Detector handles this outcome contextually (URL presence boosts score).
    // Returning null here keeps SSRS from emitting a hard verdict on its own.
    is SsrsCheckOutcome.PhishingClaimsOrgWithoutPrefix -> null
    // OFCA SSRS gates the # prefix at the carrier, so an unknown-to-us
    // prefix is still a positive signal that a real org registered it —
    // not suspicion. Let content/blocklist rules drive the verdict.
    is SsrsCheckOutcome.UnknownHashPrefix -> null
    is SsrsCheckOutcome.NoSignal -> null
}

private fun extractHashPrefix(senderId: String): String? {
    val match = HASH_PREFIX_PATTERN.find(senderId.trim()) ?: return null
    return match.groupValues[1].ifEmpty { null }
}

private fun findOrgClaimedInBody(body: String, orgs: List<OrgPrefixMapping>): OrgPrefixMapping? {
    // Only consider an "org claim" when the org name appears at the start of
    // the message (within the first 30 chars). Real impersonation looks like
    // "[ORG]:" / "ORG: ...". Mid-text mentions ("防騙：警方提醒...", "Pay your
    // HSBC card at our shop") are not impersonation and shouldn't trigger.
    val head = body.take(30)
    val lowerHead = head.lowercase()
    for (org in orgs) {
        for (alias in org.aliasesZhHk) {
            if (head.contains(alias)) return org
        }
        for (alias in org.aliasesEn) {
            if (lowerHead.contains(alias.lowercase())) return org
        }
    }
    return null
}

File: shared/rules/src/ssrs-verifier.ts
import type {
  IncomingSms,
  OrgPrefixMapping,
  SsrsRegistry,
  Verdict,
} from './types.js';

const HASH_PREFIX_PATTERN = /^#([A-Za-z0-9_]+)\b/;

export interface SsrsCheckOutcome {
  readonly outcome:
    | 'trusted_registered_prefix'
    | 'phishing_claims_org_without_prefix'
    | 'unknown_hash_prefix'
    | 'no_signal';
  readonly matchedOrg?: OrgPrefixMapping;
  readonly observedPrefix?: string;
}

export function checkSsrs(sms: IncomingSms, registry: SsrsRegistry): SsrsCheckOutcome {
  const observedPrefix = extractHashPrefix(sms.senderId);

  if (observedPrefix !== undefined) {
    const isRegistered = registry.registeredPrefixes.some(
      (p) => p.prefix.toLowerCase() === observedPrefix.toLowerCase(),
    );
    if (isRegistered) {
      return { outcome: 'trusted_registered_prefix', observedPrefix };
    }
    return { outcome: 'unknown_hash_prefix', observedPrefix };
  }

  const matchedOrg = findOrgClaimedInBody(sms.body, registry.orgToPrefix);
  if (matchedOrg !== undefined) {
    return { outcome: 'phishing_claims_org_without_prefix', matchedOrg };
  }

  return { outcome: 'no_signal' };
}

export function ssrsCheckToVerdict(outcome: SsrsCheckOutcome): Verdict | null {
  switch (outcome.outcome) {
    case 'trusted_registered_prefix':
      return {
        label: 'trusted',
        score: 1.0,
        firedRuleIds: ['ssrs.trusted_registered_prefix'],
        explanationKey: 'ssrs.trusted',
        explanationParams: { prefix: outcome.observedPrefix ?? '' },
      };
    case 'phishing_claims_org_without_prefix':
      return {
        label: 'high_confidence_phishing',
        score: 0.95,
        firedRuleIds: ['ssrs.org_claimed_without_prefix'],
        explanationKey: 'ssrs.org_claimed_without_prefix',
        explanationParams: {
          org: outcome.matchedOrg?.canonicalName ?? '',
          expected_prefix: outcome.matchedOrg?.expectedPrefix ?? '',
        },
      };
    case 'unknown_hash_prefix':
      return {
        label: 'suspicious',
        score: 0.4,
        firedRuleIds: ['ssrs.unknown_hash_prefix'],
        explanationKey: 'ssrs.unknown_hash_prefix',
        explanationParams: { prefix: outcome.observedPrefix ?? '' },
      };
    case 'no_signal':
      return null;
  }
}

function extractHashPrefix(senderId: string): string | undefined {
  const match = HASH_PREFIX_PATTERN.exec(senderId.trim());
  if (match === null || match[1] === undefined) {
    return undefined;
  }
  return match[1];
}

function findOrgClaimedInBody(
  body: string,
  orgs: readonly OrgPrefixMapping[],
): OrgPrefixMapping | undefined {
  const lowerBody = body.toLowerCase();
  for (const org of orgs) {
    for (const alias of org.aliasesZhHk) {
      if (body.includes(alias)) return org;
    }
    for (const alias of org.aliasesEn) {
      if (lowerBody.includes(alias.toLowerCase())) return org;
    }
  }
  return undefined;
}
